File: Java/Rotate-List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    /**
     * @param head: the List
     * @param k: rotate to the right k places
     * @return: the list after rotation
     */
    private int getLength(ListNode head) {
        int length = 0;
        while (head != null) {
            length ++;
            head = head.next;
        }
        return length;
    }
    public ListNode rotateRight(ListNode head, int k) {
        // write your code here
        if (head == null) return null;
        int len = getLength(head);
        if (k % len == 0) return head;
        k = k % len;
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode cur = dummy;
        for (int i = 0; i < k; i ++) {
        	cur = cur.next;
        }
        ListNode slow = dummy;
        while (cur.next != null) {
        	cur = cur.next;
        	slow = slow.next;
        }
        cur.next = dummy.next;
        dummy.next = slow.next;
        slow.next = null;
        return dummy.next;
    }
}

File: Cpp/Rotate-List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    /**
     * @param head: the list
     * @param k: rotate to the right k places
     * @return: the list after rotation
     */
    ListNode *rotateRight(ListNode *head, int k) {
        // write your code here
        if (k == 0 || head == NULL) return head;
        int num = 0;
        ListNode *node = head;
        ListNode *prev = new ListNode(0);
        while (node != NULL) {
            prev->next = node;
            prev = prev->next;
            node = node->next;
            num += 1;
        }
        if (num == 1 || k % num == 0) return head;
        int cnt = num - (k % num) - 1;
        node = head;
        for (int i = 0; i < cnt; i ++){
            node = node->next;
        }
        ListNode *node1 = node->next;
        node->next = NULL;
        prev->next = head;
        return node1;
    }
};
